spring-cloud-dataflow-completion/src/main/java/org/springframework/cloud/dataflow/completion/StreamCompletionProvider.java

```diff
@@ -35,6 +39,15 @@
 	@Autowired
 	private List<? extends ExpansionStrategy> completionExpansionStrategies = new ArrayList<>();
 
+	@PostConstruct
+	public void sortStrategies() {
+		// Here we mean HIGHEST_PRECEDENCE to win over others, so must be *last*
+		AnnotationAwareOrderComparator.sort(completionExpansionStrategies);
```

Comment:
Gotcha (and agreed))
Le 19 déc. 2015 2:38 PM, "Mark Fisher" <notifications@github.com> a écrit :

> In
> spring-cloud-dataflow-completion/src/main/java/org/springframework/cloud/dataflow/completion/StreamCompletionProvider.java
> <https://github.com/spring-cloud/spring-cloud-dataflow/pull/263#discussion_r48092421>
> :
>
> > @@ -35,6 +39,15 @@
> >  	@Autowired
> >  	private List<? extends ExpansionStrategy> completionExpansionStrategies = new ArrayList<>();
> >
> > +	@PostConstruct
> > +	public void sortStrategies() {
> > +		// Here we mean HIGHEST_PRECEDENCE to win over others, so must be *last*
> > +		AnnotationAwareOrderComparator.sort(completionExpansionStrategies);
>
> True, the culprit is not field-injection per se but autowiring
> (field-injection having no other option), *and* specifically autowiring
> of a list of otherwise arbitrarily defined beans in the context. @Ordered
> is specifically designed for lifecycle operations (it's not used for order
> of things like converters, etc). The intention with @Ordered is that its
> precedence order applies for the startup process and is reversed for the
> shutdown process, and it's also common to have multiple instances within a
> "tier" (all beans having the same order are processed within the same tier,
> where order is once again arbitrary). To me, the red flag with this
> particular PR is that you set HIGHEST_PRECEDENCE (which makes sense
> semantically) but then sort in reverse, whereas the intended usage
> scenarios treat HIGHEST_PRECEDENCE as *first* in the list and
> LOWEST_PRECEDENCE as last. That is why I used the word "abuse" .
>
> It seems as though some central component needs to know the order here,
> rather than each instance itself having an order value. That would provide
> a strictly ordered list with, for example, exactly one instance in the
> final position since it avoids the possibility of "tiers" (what happens
> when another instance is added and also has HIGHEST_PRECEDENCE for its
> order property?). It could be StreamCompletionProvider itself, but if
> that doesn't feel like its responsibility, then the next option would be a
> CompositeExpansionStrategy that manages a list internally but is injected
> as the sole ExpansionStrategy instance for the StreamCompletionProvider
> (a common pattern within spring-cloud-stream and spring-cloud-dataflow
> already). Or, if it feels like the order should be a purely
> configuration-side responsibility, we could switch to *non*-autowired
> constructor injection a nd expli citly provide the list of beans within a
> @Configuration class.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spring-cloud/spring-cloud-dataflow/pull/263/files#r48092421>
> .
>